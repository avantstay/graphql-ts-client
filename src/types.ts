export type Maybe<T> = null | undefined | T
export type Defined<T> = Exclude<T, undefined>

export type ResponseData = {
  data: any
  warnings: any
  headers: any
  status?: number
  errors: {
    message: string
  }[]
}

export type ResponseListenerInfo = {
  queryName: string
  query: string
  variables: any
  response: ResponseData
}
export type IResponseListener = (info: ResponseListenerInfo) => void | Promise<void>

type ArrayElement<ArrayType extends readonly unknown[]> = ArrayType extends readonly (infer ElementType)[] ? ElementType : never

type Primitive = Date | string | number | boolean | null | undefined

// Projection is the resulting type of Selection (type generated out of a query) applied to Base (generated graphql type)
export type Projection<Selection, Base, E = never> = Base extends Array<any>
  ? ArrayElement<Base> extends Primitive | E
    ? ArrayElement<Base>[]
    : Projection<Defined<Selection>, ArrayElement<Base>, E>[]
  : Base extends Primitive | E
  ? // Is primitive and extends undefined
    Selection extends undefined
    ? Base | undefined
    : Base
  : {
      [k in keyof Selection & keyof Base]: Selection[k] extends boolean
        ? Base[k]
        : Base[k] extends Array<infer A>
        ? Projection<Defined<Selection[k]>, A, E>[]
        : Projection<Defined<Selection[k]>, Base[k], E>
    }

export type Unpacked<T> = T extends (infer U)[]
  ? U
  : T extends (...args: any[]) => infer U
  ? U
  : T extends Promise<infer U>
  ? U
  : T

export type Replacement<M extends [any, any], T> = M extends any ? ([T] extends [M[0]] ? M[1] : never) : never

export type DeepReplace<T, Ignore, M extends [any, any]> = {
  [P in keyof T]: T[P] extends M[0]
    ? T[P] extends Ignore
      ? T[P] extends object
        ? DeepReplace<T[P], Ignore, M>
        : T[P]
      : Replacement<M, T[P]>
    : T[P] extends object
    ? DeepReplace<T[P], Ignore, M>
    : T[P]
}

export type RawEndpoint<I, O, E> = <S extends I>(
  jsonQuery?: S
) => Promise<{ data: Projection<S, O, E>; errors: any[]; warnings: any[]; headers: any; status: any }>

export type JsonOutput<O, ToBeIgnored> = DeepReplace<O, ToBeIgnored, [string | Date, string]>

export type Endpoint<I, O, E> = (<S extends I>(jsonQuery?: S) => Promise<Projection<S, JsonOutput<O, E>, E>>) & {
  memo: <S extends I>(jsonQuery?: S) => Promise<Projection<S, JsonOutput<O, E>, E>>
  memoRaw: RawEndpoint<I, JsonOutput<O, E>, E>
  raw: RawEndpoint<I, JsonOutput<O, E>, E>
}

export type ClientConfig = {
  url: string
  headers: {
    [key: string]: string
  }
  retryConfig: {
    max: number
    waitBeforeRetry?: number
    before: IResponseListener
  }
}

export type LogInfo = {
  query: string
  variables: any
  formatGraphQL: any
  kind: string
  queryName: string
  response?: any
  error?: Error
  duration: number
}

export class GraphQLClientError extends Error {
  responseData: ResponseData

  constructor(responseData: ResponseData) {
    super()
    this.responseData = responseData
    Object.setPrototypeOf(this, GraphQLClientError.prototype)
  }

  get message(): string {
    return this.response.errors.map(it => it.message).join(';\n')
  }

  get response(): ResponseData {
    return this.responseData
  }
}
