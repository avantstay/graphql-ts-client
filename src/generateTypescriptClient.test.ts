import { ApolloServer } from 'apollo-server'
import * as path from 'path'
import { generateTypescriptClient, generateTypescriptClientFromSDL } from './generateTypescriptClient'
import { startServer } from './testServer'
import { GraphQLClientError, ResponseListenerInfo } from './types'

let testServer: { server: ApolloServer; url: string }
let client: any

describe('Generated Client', () => {
  beforeAll(async () => {
    testServer = await startServer()

    const clientName = 'myApiClient'

    const { js } = await generateTypescriptClient({
      clientName,
      endpoint: `${testServer.url}graphql`,
      // For the sake of checking the generated code, we'll
      // specify an output path
      output: path.resolve(__dirname, './__testClient.ts'),
      formatGraphQL: true,
      skipCache: true,
    })

    client = eval(`${js};${clientName}`)
  })

  afterAll(async () => await testServer.server.stop())

  it('should be able to pass custom headers to a query without args', async () => {
    const books = await client.queries.booksWithoutParams({
      __headers: { 'X-Custom-Header': 'Foo' },
      title: true,
      author: true,
    })

    expect(books).toHaveLength(2)
    expect(books[0]).toHaveProperty('title')
    expect(books[0]).toHaveProperty('author')
  })

  it('should be able to make queries with optional args, not passing args obj', async () => {
    // noinspection TypeScriptValidateJSTypes
    const books = await client.queries.booksWithOptionalParams({
      title: true,
      author: true,
    })

    expect(books).toHaveLength(2)
    expect(books[0]).toHaveProperty('title')
    expect(books[0]).toHaveProperty('author')
  })

  it('should be able to make queries with optional args, not passing args obj', async () => {
    // noinspection TypeScriptValidateJSTypes
    const books = await client.queries.booksWithOptionalParams({
      __alias: 'helloWorld',
      title: true,
      author: true,
    })

    expect(books).toHaveLength(2)
    expect(books[0]).toHaveProperty('title')
    expect(books[0]).toHaveProperty('author')
  })

  it('fail with broken queries', async () => {
    const result = await client.queries
      .failingQuery({
        __args: {
          id: 'hello',
        },
      })
      .then(
        () => 'success',
        (err: any) => err
      )

    expect(result).toBeInstanceOf(GraphQLClientError)
    expect(result.message).toBe('Failed lorem ipsum dolor')
  })

  it('does not fail with broken queries when using raw requests', async () => {
    const result = await client.queries.failingQuery.raw({
      __args: {
        id: 'hello',
      },
    })

    expect(result.status).toBe(200)
    expect(result.errors).toHaveLength(1)
    expect(result.errors[0].message).toBe('Failed lorem ipsum dolor')
  })

  it('failing operations throw errors', async () => {
    let failed = false

    try {
      await client.queries.failingQuery({
        __args: {
          id: 'hello',
        },
      })
    } catch (err) {
      failed = true
    }

    expect(failed).toBe(true)
  })

  it('failing operations throw errors', async () => {
    let responseData: ResponseListenerInfo | undefined
    client.addResponseListener((data: any) => (responseData = data))

    try {
      await client.queries.failingQuery({
        __args: {
          id: 'hello',
        },
      })
    } catch {}

    expect(responseData?.queryName).toBe('failingQuery')
    expect(responseData?.response.errors.length).toBeGreaterThan(0)
  })

  it('should generate proper code from SDL', ()=>{
    const sdlString = `
  type Query {
    hello: String
  }
`;
    expect(generateTypescriptClientFromSDL(sdlString, {endpoint: 'https://sample.endpoint.com/graphl'})).toMatchSnapshot()
  })
})
