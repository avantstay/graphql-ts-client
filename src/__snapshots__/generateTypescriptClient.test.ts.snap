// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generated Client should generate proper code from SDL 1`] = `
{
  "js": "var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  __markAsModule(target);
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toModule = (module2) => {
  return __reExport(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", module2 && module2.__esModule && "default" in module2 ? { get: () => module2.default, enumerable: true } : { value: module2, enumerable: true })), module2);
};
__export(exports, {
  client: () => client,
  default: () => stdin_default
});
var import_endpoint = __toModule(require("./endpoint"));
const formatGraphQL = (query) => query;
const typesTree = {};
let verbose = false;
let headers = {};
let url = "https://sample.endpoint.com/graphl";
let retryConfig = {
  max: 0,
  before: void 0,
  waitBeforeRetry: 0
};
let responseListeners = [];
let errorsParser = void 0;
let apiEndpoint = (0, import_endpoint.getApiEndpointCreator)({
  getClient: () => ({ url, headers, retryConfig }),
  responseListeners,
  maxAge: 3e4,
  verbose,
  typesTree,
  formatGraphQL,
  errorsParser
});
const client = {
  addResponseListener: (listener) => responseListeners.push(listener),
  setHeader: (key, value) => {
    headers[key] = value;
  },
  setHeaders: (newHeaders) => {
    headers = newHeaders;
  },
  setRetryConfig: (options) => {
    if (!Number.isInteger(options.max) || options.max < 0) {
      throw new Error("retryOptions.max should be a non-negative integer");
    }
    retryConfig = {
      max: options.max,
      waitBeforeRetry: options.waitBeforeRetry,
      before: options.before
    };
  },
  setUrl: (_url) => url = _url,
  queries: {
    hello: apiEndpoint("query", "hello")
  },
  mutations: {}
};
var stdin_default = client;
",
  "typings": "// noinspection TypeScriptUnresolvedVariable, ES6UnusedImports, JSUnusedLocalSymbols, TypeScriptCheckImport
import { DeepRequired } from "ts-essentials"
import { Maybe, IResponseListener, Endpoint } from "."

// Scalars
export type IDate = string | Date

// Enums

type AllEnums = never

// Args
export interface HelloArgs {}

// Input/Output Types

/**
 * @deprecated Avoid directly using this interface. Instead, create a type alias based on the query/mutation return type.
 */

export interface Query {
  hello?: string
}

// Selection Types

export declare const client: {
  addResponseListener: (listener: IResponseListener) => void
  setHeader: (key: string, value: string) => void
  setHeaders: (newHeaders: { [k: string]: string }) => void
  setUrl: (url: string) => void
  setRetryConfig: (options: {
    max: number
    waitBeforeRetry?: number
    before?: IResponseListener
  }) => void
  queries: {
    hello: Endpoint<
      {
        __headers?: { [key: string]: string }
        __retry?: boolean
        __alias?: string
        __url?: string
      },
      string,
      AllEnums
    >
  }
  mutations: {}
}

export default client
",
}
`;
